### PR TITLE
[Enhance] Call `register_all_modules`  in `Registry.get()`

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

If we use config like this:

```python
model = dict(type='RetinaNet'
    backbone=dict(type='mmcls.ResNet'
    ...
    )
)
```

`MODELS.build(model)` will **NOT** regist modules of  `mmcl`, and we need to set `custom_configs=...` in config file extraly.

However, if we set `_scope_` in config:

```python
model = dict(type='RetinaNet'
    _scope_='mmcls',
    backbone=dict(type='ResNet'
    ...
    )
)
```
The modules of `mmcls` will be registered, and its behavior is inconsistent with `type=mmcls.ResNet`.
Therefore, this pr make `type=mmcls.ResNet` will also trigger the regist process.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by Harold-lkk — APPROVED** [review]

File: mmengine/registry/registry.py [hunk 1]
```diff
@@ -243,7 +243,7 @@ def switch_scope_and_registry(self, scope: str) -> Generator:
                 if scope_name in PKG2PROJECT:
                     try:
                         module = import_module(f'{scope_name}.utils')
-                        module.register_all_modules()  # type: ignore
+                        module.register_all_modules(False)  # type: ignore
                     except ImportError as e:
                         raise e
```
Comment (Harold-lkk) [inline mmengine/registry/registry.py:248]: Better adding `print_log` too

File: mmengine/registry/registry.py [hunk 2]
```diff
@@ -347,6 +347,14 @@ def get(self, key: str) -> Optional[Type]:
...
                     parent = parent.parent
         else:
+            try:
+                module = import_module(f'{scope_name}.utils')
+                module.register_all_modules(False)  # type: ignore
+            except ImportError:
+                print_log(f'{scope_name} is not a package of OpenMMLab, '
+                          'and its modules will not be registered. If you '
+                          f'want to use modules defined in {scope_name}, '
+                          f'please import corresponding modules firstly.')
```
Comment (Harold-lkk) [inline mmengine/registry/registry.py:357]: or remind user to pip install {scope_name}

File: mmengine/registry/registry.py [hunk 3]
```diff
@@ -347,6 +347,14 @@ def get(self, key: str) -> Optional[Type]:
                         break
                     parent = parent.parent
         else:
+            try:
+                module = import_module(f'{scope_name}.utils')
```
Comment (fangyixiao18) [inline mmengine/registry/registry.py:351]:
```suggestion
                module = import_module(f'{scope}.utils')
```

**Review by fangyixiao18 — DISMISSED** [review]

File: mmengine/registry/registry.py [hunk 4]
```diff
@@ -243,9 +243,24 @@ def switch_scope_and_registry(self, scope: str) -> Generator:
...
+                                'want to use modules defined in '
+                                f'{scope}, Please install {scope} by '
+                                f'`pip install {PKG2PROJECT[scope]}.',
+                                logger='current',
+                                level=logging.WARNING)
+                        else:
+                            print_log(
+                                f'Failed to import {scope} and register '
+                                'its modules, please make sure you '
+                                'haveregister the module mannuly.',
```
Comment (zhouzaida) [inline mmengine/registry/registry.py:261]:
```suggestion
                                'have registered the module mannuly.',
```

**Review by zhouzaida — DISMISSED** [review]

**Review by Harold-lkk — DISMISSED** [review]

**Review by RangiLyu — APPROVED** [review]

File: mmengine/registry/registry.py [hunk 5]
```diff
@@ -243,9 +243,24 @@ def switch_scope_and_registry(self, scope: str) -> Generator:
...
+                                'want to use modules defined in '
+                                f'{scope}, Please install {scope} by '
+                                f'`pip install {PKG2PROJECT[scope]}.',
+                                logger='current',
+                                level=logging.WARNING)
+                        else:
+                            print_log(
+                                f'Failed to import {scope} and register '
+                                'its modules, please make sure you '
+                                'have registered the module mannuly.',
```
Comment (RangiLyu) [inline mmengine/registry/registry.py:261]:
```suggestion
                                'have registered the module manually.',
```

File: mmengine/registry/registry.py [hunk 6]
```diff
@@ -347,6 +362,24 @@ def get(self, key: str) -> Optional[Type]:
...
+                        f'{scope} is not installed and its modules '
+                        'will not be registered. If you want to use '
+                        f'modules defined in {scope}, Please install '
+                        f'{scope} by `pip install {PKG2PROJECT[scope]} ',
+                        logger='current',
+                        level=logging.WARNING)
+                else:
+                    print_log(
+                        f'Failed to import "{scope}", and register its '
+                        f'modules. Please register {real_key} mannuly.',
```
Comment (RangiLyu) [inline mmengine/registry/registry.py:380]:
```suggestion
                        f'modules. Please register {real_key} manually.',
```

**Review by RangiLyu — COMMENTED** [review]
Please fix the typos

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**